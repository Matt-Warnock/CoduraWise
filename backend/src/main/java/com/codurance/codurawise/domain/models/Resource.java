package com.codurance.codurawise.domain.models;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Objects;

public class Resource {

  private Integer id;
  private String title;
  private String link;
  private Double averageRating;
  private String mediaType;
  private Collection<Tag> tags;

  public static Resource of(Integer id, String title, String link, Double averageRating, String mediaType) {
    Resource resource = new Resource();
    resource.setId(id);
    resource.setTitle(title);
    resource.setLink(link);
    resource.setAverageRating(averageRating);
    resource.setMediaType(mediaType);
    return resource;
  }

  public static Resource of(Integer id, String title, String link, Double averageRating, String mediaType, Collection<Tag> tags) {
    Resource resource = of(id, title, link, averageRating, mediaType);
    resource.setTags(tags);
    return resource;
  }

  public Integer getId() {
    return id;
  }

  public void setId(Integer id) {
    this.id = id;
  }

  public void setTitle(String title) {
    this.title = title;
  }

  public String getTitle() {
    return title;
  }

  public void setLink(String link) {
    this.link = link;
  }

  public String getLink() {
    return link;
  }

  public void setAverageRating(Double averageRating) {
    this.averageRating = averageRating;
  }

  public Double getAverageRating() {
    return averageRating;
  }

  public void setMediaType(String mediaType) {
    this.mediaType = mediaType;
  }

  public String getMediaType() {
    return mediaType;
  }

  public void setTags(Collection<Tag> tags) {
    this.tags = new ArrayList<>(tags);
  }

  public Collection<Tag> getTags() {
    return Collections.unmodifiableCollection(tags);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    Resource resource = (Resource) o;

    if (!Objects.equals(id, resource.id)) return false;
    if (!Objects.equals(title, resource.title)) return false;
    if (!Objects.equals(link, resource.link)) return false;
    if (!Objects.equals(averageRating, resource.averageRating))
      return false;
    if (!Objects.equals(mediaType, resource.mediaType)) return false;
    return Objects.equals(tags, resource.tags);
  }

  @Override
  public int hashCode() {
    int result = id != null ? id.hashCode() : 0;
    result = 31 * result + (title != null ? title.hashCode() : 0);
    result = 31 * result + (link != null ? link.hashCode() : 0);
    result = 31 * result + (averageRating != null ? averageRating.hashCode() : 0);
    result = 31 * result + (mediaType != null ? mediaType.hashCode() : 0);
    result = 31 * result + (tags != null ? tags.hashCode() : 0);
    return result;
  }

  @Override
  public String toString() {
    return "Resource{" +
      "id=" + id +
      ", title='" + title + '\'' +
      ", link='" + link + '\'' +
      ", averageRating=" + averageRating +
      ", mediaType='" + mediaType + '\'' +
      ", tags=" + tags +
      '}';
  }
}
