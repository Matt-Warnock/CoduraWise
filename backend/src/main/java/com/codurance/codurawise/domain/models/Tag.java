package com.codurance.codurawise.domain.models;

import java.util.Objects;

public class Tag {

  private String tag;

  public static Tag of(String tagValue) {
    Tag tag = new Tag();
    tag.setTag(tagValue);
    return tag;
  }

  public String getTag() {
    return tag;
  }

  public void setTag(String tag) {
    this.tag = tag;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    Tag tag1 = (Tag) o;

    return Objects.equals(tag, tag1.tag);
  }

  @Override
  public int hashCode() {
    return tag != null ? tag.hashCode() : 0;
  }

  @Override
  public String toString() {
    return "Tag{" +
      "tag='" + tag + '\'' +
      '}';
  }
}
