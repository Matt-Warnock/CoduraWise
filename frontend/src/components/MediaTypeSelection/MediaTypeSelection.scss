.media-selection {
  width: 10em;
  margin: 10px;

  border: none;
  padding: 1em;

  font-family: monospace, sans-serif;
  font-size: 0.9em;

  & legend {
    font-size: 1em;
    margin: 5px;
  }
  & input {
    margin-left: 5px;
  }
  & label {
    margin-left: 5px;
  }
}
