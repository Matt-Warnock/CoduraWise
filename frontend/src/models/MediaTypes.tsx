export type MediaType = "video" | "article" | "book" | "tutorial" | "course";

export const MediaTypes: Array<MediaType> = [
  "video", "article", "book", "tutorial", "course"
];