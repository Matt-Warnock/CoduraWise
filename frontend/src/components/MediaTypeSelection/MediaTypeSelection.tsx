import React from "react";
import { Dispatch, SetStateAction, useState } from "react";
import { MediaType, MediaTypes } from "../../models/MediaTypes";
import "./MediaTypeSelection.scss";

interface MediaTypeSelectionProps {
  setFilterMediaTypes: Dispatch<SetStateAction<MediaType[]>>;
}

const MediaTypeSelection = ({
  setFilterMediaTypes,
}: MediaTypeSelectionProps) => {
  const [checkedState, setCheckedState] = useState(
    new Array(MediaTypes.length).fill(false),
  );

  const createFilter = (checkedArray: Array<boolean>) => {
    const mediaTypesChecked: Array<MediaType> = [];
    checkedArray.map((value: boolean, index: number) => {
      if (value) {
        mediaTypesChecked.push(MediaTypes[index]);
      }
    });
    return mediaTypesChecked;
  };

  const handleOnChange = (mediaType: MediaType, position: number) => {
    // create new state with the checked value
    const updatedCheckedState = checkedState.map((item, index) =>
      index === position ? !item : item,
    );
    // create MediaType array from check list
    const mediaTypesChecked = createFilter(updatedCheckedState);
    // set the context
    setFilterMediaTypes(mediaTypesChecked);
    // set the state
    setCheckedState(updatedCheckedState);
  };

  return (
    <form className="media-selection">
      <legend className="media-selection__legend">Filter by</legend>
      {MediaTypes.map((mediaType: MediaType, index) => (
        <div key={mediaType}>
          <input
            className="media-selection__input"
            type="checkbox"
            name={mediaType}
            id={`${mediaType}media`}
            value={mediaType}
            onChange={() => handleOnChange(mediaType, index)}
            checked={checkedState[index]}
          />
          <label
            className="media-selection__label"
            htmlFor={`${mediaType}media`}
          >
            {mediaType}
          </label>
        </div>
      ))}
    </form>
  );
};

export default MediaTypeSelection;
