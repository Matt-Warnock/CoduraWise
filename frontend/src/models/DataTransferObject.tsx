import { MediaType } from "./MediaTypes";

export interface AddResourceBodyDTO {
  title: string;
  link: string;
  tags: Array<string>;
  averageRating: string;
  mediaType: MediaType;
}

