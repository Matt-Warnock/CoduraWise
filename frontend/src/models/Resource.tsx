import { MediaType } from "./MediaTypes";

export interface Resource {
  id: number;
  title: string;
  link: string;
  mediaType: MediaType;
  averageRating: number;
}
