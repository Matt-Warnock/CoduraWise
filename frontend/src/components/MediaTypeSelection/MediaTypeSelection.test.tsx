import TestRenderer from "react-test-renderer";
import MediaTypeSelection from "./MediaTypeSelection";
import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { MediaType } from "../../models/MediaTypes";

test("snapshot test: the html matches", () => {
  const card = TestRenderer.create(
    <MediaTypeSelection setFilterMediaTypes={jest.fn()} />,
  );

  expect(card.toJSON()).toMatchSnapshot();
});

describe("Given a media filter selection component", () => {
  let mockSetFilterMediaTypes = jest.fn();
  beforeEach(() => {
    mockSetFilterMediaTypes = jest.fn();
  });

  describe("When I check the checkboxes of video and book", () => {
    test("Then it updates the state to only show videos and books", () => {
      render(
        <MediaTypeSelection setFilterMediaTypes={mockSetFilterMediaTypes} />,
      );

      const videoCheckbox = screen.getByRole("checkbox", { name: "video" });
      const bookCheckBox = screen.getByRole("checkbox", { name: "book" });

      fireEvent.click(videoCheckbox);
      fireEvent.click(bookCheckBox);

      const expectedArrayVideo: Array<MediaType> = ["video"];
      const expectedArrayVideoAndBook: Array<MediaType> = ["video", "book"];

      expect(mockSetFilterMediaTypes).toHaveBeenNthCalledWith(
        1,
        expectedArrayVideo,
      );
      expect(mockSetFilterMediaTypes).toHaveBeenNthCalledWith(
        2,
        expectedArrayVideoAndBook,
      );
    });
  });

  describe("When I check the book checkbox and then uncheck the book checkbox", () => {
    test("Then it updates the state to show all resources", () => {
      render(
        <MediaTypeSelection setFilterMediaTypes={mockSetFilterMediaTypes} />,
      );

      const bookCheckBox = screen.getByRole("checkbox", { name: "book" });
      fireEvent.click(bookCheckBox);
      fireEvent.click(bookCheckBox);

      const expectedArrayBook: Array<MediaType> = ["book"];
      const expectedArrayNoFilterSelected: Array<MediaType> = [];

      expect(mockSetFilterMediaTypes).toHaveBeenNthCalledWith(
        1,
        expectedArrayBook,
      );
      expect(mockSetFilterMediaTypes).toHaveBeenNthCalledWith(
        2,
        expectedArrayNoFilterSelected,
      );
    });
  });
});

