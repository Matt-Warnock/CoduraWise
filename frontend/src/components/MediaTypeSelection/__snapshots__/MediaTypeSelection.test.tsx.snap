// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: the html matches 1`] = `
<form
  className="media-selection"
>
  <legend
    className="media-selection__legend"
  >
    Filter by
  </legend>
  <div>
    <input
      checked={false}
      className="media-selection__input"
      id="videomedia"
      name="video"
      onChange={[Function]}
      type="checkbox"
      value="video"
    />
    <label
      className="media-selection__label"
      htmlFor="videomedia"
    >
      video
    </label>
  </div>
  <div>
    <input
      checked={false}
      className="media-selection__input"
      id="articlemedia"
      name="article"
      onChange={[Function]}
      type="checkbox"
      value="article"
    />
    <label
      className="media-selection__label"
      htmlFor="articlemedia"
    >
      article
    </label>
  </div>
  <div>
    <input
      checked={false}
      className="media-selection__input"
      id="bookmedia"
      name="book"
      onChange={[Function]}
      type="checkbox"
      value="book"
    />
    <label
      className="media-selection__label"
      htmlFor="bookmedia"
    >
      book
    </label>
  </div>
  <div>
    <input
      checked={false}
      className="media-selection__input"
      id="tutorialmedia"
      name="tutorial"
      onChange={[Function]}
      type="checkbox"
      value="tutorial"
    />
    <label
      className="media-selection__label"
      htmlFor="tutorialmedia"
    >
      tutorial
    </label>
  </div>
  <div>
    <input
      checked={false}
      className="media-selection__input"
      id="coursemedia"
      name="course"
      onChange={[Function]}
      type="checkbox"
      value="course"
    />
    <label
      className="media-selection__label"
      htmlFor="coursemedia"
    >
      course
    </label>
  </div>
</form>
`;
